torch/optim/_muon.py

```diff
@@ -0,0 +1,276 @@
+DEFAULT_C = 2.0315
+
+
+@dataclass
+class BaseMsignFnConfig:
+    """Configuration used by :func:`msign_fn`."""
+
+
+@dataclass
+class NewtonSchulzConfig(BaseMsignFnConfig):
+    # """Configuration used by :func:`zeropower_via_newtonschulz`."""
+
+    coefficients: tuple[float, float, float] = (DEFAULT_A, DEFAULT_B, DEFAULT_C)
+    eps: float = EPS
+    ns_steps: int = 5
+
+
+AdjustLrFn: TypeAlias = Callable[[float, torch.Size], float]
+MsignFn: TypeAlias = Callable[[Tensor, BaseMsignFnConfig], Tensor]
+
+
+def zeropower_via_newtonschulz(grad: Tensor, ns_config: BaseMsignFnConfig) -> Tensor:
+    """
+    Newton-Schulz iteration to compute the zeroth power / orthogonalization of G. We opt to use a
+    quintic iteration whose coefficients are selected to maximize the slope at zero. For the purpose
+    of minimizing steps, it turns out to be empirically effective to keep increasing the slope at
+    zero even beyond the point where the iteration no longer converges all the way to one everywhere
+    on the interval. This iteration therefore does not produce UV^T but rather something like US'V^T
+    where S' is diagonal with S_{ii}' ~ Uniform(0.5, 1.5), which turns out not to hurt model
+    performance at all relative to UV^T, where USV^T = G is the SVD.
+
+    Implementation reference: https://github.com/KellerJordan/Muon/blob/master/muon.py
+    with suggestions by @jxbz, @leloykun, and @YouJiacheng.
+    """
+    ns_config = cast(NewtonSchulzConfig, ns_config)
+    steps = ns_config.ns_steps
+    eps = ns_config.eps
+    coefficients = ns_config.coefficients
+    assert steps < 100, (
+        "Number of steps must be less than 100 for computational efficiency"
+    )
+    assert len(grad.shape) == 2, "Input tensor gradient must be a 2D matrix"
+    assert len(coefficients) == 3, "Coefficients must be a tuple of exactly 3 values"
+    a, b, c = coefficients[0], coefficients[1], coefficients[2]
+    X = grad.bfloat16()
+    if grad.size(0) > grad.size(1):
+        X = X.T
+    # Ensure spectral norm is at most 1
+    X = X / (X.norm() + eps)
+    # Perform the NS iterations
+    for _ in range(steps):
+        A = X @ X.T
+        B = b * A + c * A @ A
+        X = a * X + B @ X
+
+    if grad.size(0) > grad.size(1):
+        X = X.T
+    return X
+
+
+def default_adjust_lr(lr: float, param_shape: torch.Size) -> float:
+    """Default learning rate adjustment used by Muon. Method reported in the paper https://arxiv.org/pdf/2502.16982."""
+    A, B = param_shape[:2]
+    adjusted_ratio = 0.2 * math.sqrt(max(A, B))
+    return lr * adjusted_ratio
+
+
+class Muon(Optimizer):
+    """Implements the Muon optimizer.
+
+    This optimizer performs momentum SGD followed by an optional orthogonalization
+    step computed via a user provided callable.
+    """
+
+    def __init__(
+        self,
+        params: ParamsT,
+        lr: float = 1e-3,
+        weight_decay: float = 0.1,
+        momentum: float = 0.95,
+        nesterov: bool = True,
```

Comment:
can you link the impl or paper from which these values are based on?

Reply:
added!

Reply:
thanks! Also, I think this can go in a following PR, but it would be very nice to have a LaTeX render of the mathematical formula for the update similar to what we have in some other optims like https://docs.pytorch.org/docs/stable/generated/torch.optim.AdamW.html esp if we see Muon as the next big optimizer